src/osd/PrimaryLogPG.cc

```diff
@@ -5286,11 +5286,11 @@ struct FillInVerifyExtent : public Context {
     r(r), rval(rv), outdatap(blp), maybe_crc(mc),
     size(size), osd(osd), soid(soid), flags(flags) {}
   void finish(int len) override {
-    *r = len;
     if (len < 0) {
       *rval = len;
       return;
     }
+    *r = len;
```

Comment:
could you please explain the impact of this issue? how / why does this change help?

Reply:
When read failed, ret can be taken as data len in FillInVerifyExtent, which should be avoided.

Reply:
@yanqiang-ux could you elaborate the consequence of the issue? so we can evaluate the severity of it. and consider if this should be backported, etc.

Reply:
It may cause errors in crc repair or retry read because of the data len. In my case, we use FillInVerifyExtent for EC read，when meet -EIO，we will try crc repair, which need read data from other shard accrding to data len. and I meet assert in   ECBackend.cc (loc:  line 2288  ceph_assert(range.first != range.second) ), But it seems master branch not support EC crc repair.
In shot, when reuse the readop may cause unpredictable error.

Reply:
@yanqiang-ux could you update the commit message with the analysis ?

Reply:
plese note the first comment in a PR is not the same thing as the commit message. please use 

```
git commit --amend
```

for updating your commit message.

Reply:
Updated

Reply:
so you are referencing the call in `PrimaryLogPG::do_read()`, where `op.extent.length` is used as both the input parameter **and** the output parameter. so if `FillInVerifyExtent` is called with a negative error no, we end up with a negative `op.extent.length`. this leaves us a corrupted `op`. and we cannot use `op` anymore even after repairing the object being accessed.